[Router] Refactor protocol definitions: split spec.rs into modular files

key4ng opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation
The spec.rs file had grown to 2739 lines, making it difficult to maintain and navigate. This violates the code style guideline of splitting files over 2000 lines. A modular structure improves code organization, makes it easier to find specific protocol definitions, and reduces merge conflicts when multiple developers work on different API endpoints.
<!-- Describe the purpose and goals of this pull request. -->

## Modifications
1. Split spec.rs into 7 modular files organized by API type:
    - chat.rs - Chat completions API (668 lines)
    - completion.rs - Completions API (213 lines)
    - generate.rs - Generate API (288 lines)
    - embedding.rs - Embeddings API (57 lines)
    - rerank.rs - Rerank API (215 lines)
    - responses.rs - OpenAI Responses API (912 lines)
    - common.rs - Shared types and utilities (358 lines)
  2. Consolidated shared utilities in common.rs:
    - Moved default_model() and default_true() helper functions
    - Moved GenerationRequest trait for unified request handling
    - Kept common types like UsageInfo, StringOrArray, SamplingParams
  3. Improved validation consistency:
    - Refactored RerankRequest to use validator crate's #[derive(Validate)] pattern
    - Added custom validation functions (validate_query, validate_documents, validate_rerank_request)
    - Made validation approach consistent across all request types
  4. Code cleanup:
    - Removed test-only helper UsageInfo::new_with_cached() from production code
    - Removed unused _sampling_params parameter from ResponsesResponse::from_request()
    - Fixed regression in rerank response building
    - Updated all imports across routers, tests, and benchmarks
  5. Maintained backward compatibility: No API changes, only internal reorganization
<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->
grpc chat completions api
```
ubuntu@moirai-a10-exp:~/sglang/sgl-router$ curl -X POST http://localhost:3002/v1/chat/completions   -H "Content-Type: application/json"   -d '{
    "model": "llama-3.1-8b-instruct",
    "messages": [
      {"role": "user", "content": "What is the answer to 1 + 32920?"}
    ],
    "stream": false,
    "max_tokens": 200,
    "top_k": -1,
    "logprobs": true,
    "n": 2,
    "top_logprobs": 2
  }'
{"id":"chatcmpl-1e9acd26-a6bb-4414-ae20-a8423c00ec08","object":"chat.completion","created":1760557826,"model":"llama-3.1-8b-instruct","choices":[{"index":0,"message":{"role":"assistant","content":"1 + 32920 = 32921"},"logprobs":{"content":[{"token":"1","logprob":-0.8048144,"bytes":[49],"top_logprobs":[{"token":"1","logprob":-0.8048144,"bytes":[49]},{"token":"The","logprob":-1.1798143,"bytes":[84,104,101]}]},{"token":" +","logprob":-0.00002527269,"bytes":[32,43],"top_logprobs":[{"token":" +","logprob":-0.00002527269,"bytes":[32,43]},{"token":" (","logprob":-11.375026,"bytes":[32,40]}]},{"token":" ","logprob":-4.768373e-7,"bytes":[32],"top_logprobs":[{"token":" ","logprob":-4.768373e-7,"bytes":[32]},{"token":" (","logprob":-14.750001,"bytes":[32,40]}]},{"token":"329","logprob":-0.00021573248,"bytes":[51,50,57],"top_logprobs":[{"token":"329","logprob":-0.00021573248,"bytes":[51,50,57]},{"token":"32","logprob":-8.750216,"bytes":[51,50]}]},{"token":"20","logprob":-0.0000019073505,"bytes":[50,48],"top_logprobs":[{"token":"20","logprob":-0.0000019073505,"bytes":[50,48]},{"token":"21","logprob":-14.000002,"bytes":[50,49]}]},{"token":" =","logprob":-0.0021107404,"bytes":[32,61],"top_logprobs":[{"token":" =","logprob":-0.0021107404,"bytes":[32,61]},{"token":" is","logprob":-6.752111,"bytes":[32,105,115]}]},{"token":" ","logprob":-0.0000088215265,"bytes":[32],"top_logprobs":[{"token":" ","logprob":-0.0000088215265,"bytes":[32]},{"token":" \n\n","logprob":-11.750009,"bytes":[32,10,10]}]},{"token":"329","logprob":-0.0000013113031,"bytes":[51,50,57],"top_logprobs":[{"token":"329","logprob":-0.0000013113031,"bytes":[51,50,57]},{"token":"328","logprob":-13.875001,"bytes":[51,50,56]}]},{"token":"21","logprob":0.0,"bytes":[50,49],"top_logprobs":[{"token":"21","logprob":0.0,"bytes":[50,49]},{"token":"20","logprob":-16.75,"bytes":[50,48]}]},{"token":"<|eot_id|>","logprob":-0.57600033,"bytes":[60,124,101,111,116,95,105,100,124,62],"top_logprobs":[{"token":"<|eot_id|>","logprob":-0.57600033,"bytes":[60,124,101,111,116,95,105,100,124,62]},{"token":".","logprob":-0.82600033,"bytes":[46]}]}]},"finish_reason":"stop","matched_stop":128009},{"index":1,"message":{"role":"assistant","content":"The answer to 1 + 32920 is 32921."},"logprobs":{"content":[{"token":"The","logprob":-1.1798143,"bytes":[84,104,101],"top_logprobs":[{"token":"1","logprob":-0.8048144,"bytes":[49]},{"token":"The","logprob":-1.1798143,"bytes":[84,104,101]}]},{"token":" answer","logprob":-0.0004758775,"bytes":[32,97,110,115,119,101,114],"top_logprobs":[{"token":" answer","logprob":-0.0004758775,"bytes":[32,97,110,115,119,101,114]},{"token":" result","logprob":-8.625476,"bytes":[32,114,101,115,117,108,116]}]},{"token":" to","logprob":-0.011049775,"bytes":[32,116,111],"top_logprobs":[{"token":" to","logprob":-0.011049775,"bytes":[32,116,111]},{"token":" is","logprob":-4.5110497,"bytes":[32,105,115]}]},{"token":" ","logprob":-0.00017996262,"bytes":[32],"top_logprobs":[{"token":" ","logprob":-0.00017996262,"bytes":[32]},{"token":" the","logprob":-8.62518,"bytes":[32,116,104,101]}]},{"token":"1","logprob":-0.00003659792,"bytes":[49],"top_logprobs":[{"token":"1","logprob":-0.00003659792,"bytes":[49]},{"token":"329","logprob":-10.250036,"bytes":[51,50,57]}]},{"token":" +","logprob":-0.000013709162,"bytes":[32,43],"top_logprobs":[{"token":" +","logprob":-0.000013709162,"bytes":[32,43]},{"token":" plus","logprob":-11.250013,"bytes":[32,112,108,117,115]}]},{"token":" ","logprob":-3.5762793e-7,"bytes":[32],"top_logprobs":[{"token":" ","logprob":-3.5762793e-7,"bytes":[32]},{"token":" (","logprob":-15.375,"bytes":[32,40]}]},{"token":"329","logprob":-0.00024816493,"bytes":[51,50,57],"top_logprobs":[{"token":"329","logprob":-0.00024816493,"bytes":[51,50,57]},{"token":"32","logprob":-8.500248,"bytes":[51,50]}]},{"token":"20","logprob":-0.0000021457695,"bytes":[50,48],"top_logprobs":[{"token":"20","logprob":-0.0000021457695,"bytes":[50,48]},{"token":"21","logprob":-13.875002,"bytes":[50,49]}]},{"token":" is","logprob":-0.00061196927,"bytes":[32,105,115],"top_logprobs":[{"token":" is","logprob":-0.00061196927,"bytes":[32,105,115]},{"token":" =","logprob":-7.875612,"bytes":[32,61]}]},{"token":" ","logprob":-0.008814144,"bytes":[32],"top_logprobs":[{"token":" ","logprob":-0.008814144,"bytes":[32]},{"token":":\n\n","logprob":-4.8838143,"bytes":[58,10,10]}]},{"token":"329","logprob":-0.00000202656,"bytes":[51,50,57],"top_logprobs":[{"token":"329","logprob":-0.00000202656,"bytes":[51,50,57]},{"token":"328","logprob":-13.250002,"bytes":[51,50,56]}]},{"token":"21","logprob":-1.192093e-7,"bytes":[50,49],"top_logprobs":[{"token":"21","logprob":-1.192093e-7,"bytes":[50,49]},{"token":",","logprob":-17.0,"bytes":[44]}]},{"token":".","logprob":-0.00049412536,"bytes":[46],"top_logprobs":[{"token":".","logprob":-0.00049412536,"bytes":[46]},{"token":"<|eot_id|>","logprob":-7.625494,"bytes":[60,124,101,111,116,95,105,100,124,62]}]},{"token":"<|eot_id|>","logprob":-0.000055433855,"bytes":[60,124,101,111,116,95,105,100,124,62],"top_logprobs":[{"token":"<|eot_id|>","logprob":-0.000055433855,"bytes":[60,124,101,111,116,95,105,100,124,62]},{"token":" \n\n","logprob":-10.000055,"bytes":[32,10,10]}]}]},"finish_reason":"stop","matched_stop":128009}],"usage":{"prompt_tokens":92,"completion_tokens":25,"total_tokens":117},"system_fingerprint":"default"}
```
openai router response api
```
ubuntu@moirai-a10-exp:~/sglang/sgl-router$ curl http://localhost:30000/v1/responses -H "Content-Type: application/json" -H "Authorization: Bearer $OPENAI_API_KEY" -d '{
"model": "gpt-5-nano",
"input": "Tell me a 3 sentence bedtime story about oracle"
}' 
{"id":"resp_0bb309a72f35dadb0168effb40cfdc81969c4ae8ea1bec33f1","object":"response","created_at":1760557889,"status":"completed","background":false,"billing":{"payer":"developer"},"error":null,"incomplete_details":null,"instructions":null,"max_output_tokens":null,"max_tool_calls":null,"model":"gpt-5-nano-2025-08-07","output":[{"id":"rs_0bb309a72f35dadb0168effb4343d08196a2f42bdc700964be","type":"reasoning","summary":[]},{"id":"msg_0bb309a72f35dadb0168effb494cac8196b00ec6f50d0d36f4","type":"message","status":"completed","content":[{"type":"output_text","annotations":[],"logprobs":[],"text":"Under the silver moon, a gentle oracle lived in a hidden grove where the wind carried soft promises.  \nChildren came with sleepy eyes and whispered wishes, and the oracle answered with a map of twinkling dreams that led them safely through the night.  \nAs the first light touched the pines, the dreams faded to a quiet sigh, and the oracle tucked its lantern of stars away until tomorrow."}],"role":"assistant"}],"parallel_tool_calls":true,"previous_response_id":null,"prompt_cache_key":null,"reasoning":{"effort":"medium","summary":null},"safety_identifier":null,"service_tier":"default","store":false,"temperature":1.0,"text":{"format":{"type":"text"},"verbosity":"medium"},"tool_choice":"auto","tools":[],"top_logprobs":0,"top_p":1.0,"truncation":"disabled","usage":{"input_tokens":16,"input_tokens_details":{"cached_tokens":0},"output_tokens":854,"output_tokens_details":{"reasoning_tokens":768},"total_tokens":870},"user":null,"metadata":{}}
```

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

CatherineSue committed:
Fix error message format

key4ng committed:
remove to_sampling_params

key4ng committed:
fix default model

key4ng committed:
move generate response id logic to middleware:

key4ng committed:
fix fmt

CatherineSue committed:
split spec

key4ng committed:
fix response request id

key4ng committed:
move response api out of spec

key4ng committed:
update pd router

key4ng committed:
fix regression: copy back stream completion response

key4ng committed:
remove spec.rs and update import

key4ng committed:
fix clippy

key4ng committed:
fix tests

key4ng committed:
fix fmt

key4ng committed:
resolve merge conflict

key4ng committed:
fix import

key4ng committed:
fix rerank regression and move default_model/default_true to common

key4ng committed:
move generaterequest trait to common

key4ng commented on sgl-router/src/protocols/chat.rs:
```diff
@@ -0,0 +1,696 @@
+
+    /// Extract text content for routing decisions
+    fn extract_text_for_routing(&self) -> String;
+}
```
thanks

key4ng commented on sgl-router/src/protocols/responses.rs:
```diff
@@ -0,0 +1,930 @@
+            reasoning_tokens,
+            prompt_tokens_details: Some(PromptTokenUsageInfo { cached_tokens }),
+        }
+    }
```
moved thanks

key4ng committed:
move test only method to test

key4ng committed:
minor clean

key4ng commented:
/gemini review

key4ng commented on sgl-router/src/protocols/responses.rs:
```diff
@@ -0,0 +1,912 @@
+        .duration_since(std::time::UNIX_EPOCH)
+        .unwrap_or_else(|_| std::time::Duration::from_secs(0))
+        .as_secs() as i64
+}
```
minor

key4ng committed:
make rerank use new validator structure

key4ng committed:
resolve comments

CatherineSue commented on sgl-router/src/protocols/responses.rs:
```diff
@@ -0,0 +1,912 @@
+    }
+}
+
+fn generate_response_id() -> String {
```
Why do we keep this? I thought it was removed in #11621 ?

CatherineSue committed:
Remove export all in protocols/mod.rs

- Export all will prevent other modules from knowing which crate is
from which file

CatherineSue committed:
Fix all imports in tool_parser/parsers

CatherineSue committed:
Fix all imports in src/

CatherineSue committed:
Fix imports in tests

CatherineSue committed:
Format files

CatherineSue committed:
Fix rerank validation tests

- Since we moved validation to server.rs

key4ng committed:
resolve comments

key4ng committed:
resolve merge conflict

key4ng committed:
remove unused sort logic for rerank

key4ng commented on sgl-router/src/protocols/rerank.rs:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    /// Sort results by score in descending order
+    pub fn sort_by_score(&mut self) {
```
You are correct, the python worker already has the sort logic, it's duplicated here. I just removed
https://github.com/sgl-project/sglang/blob/85ebeecf0694bdb94ace25319775d4b060baf655/python/sglang/srt/entrypoints/openai/serving_rerank.py#L102

key4ng replied:
I'll keep the create_timestamp method for now, since the python response doesn't return the `created` field. ideally it should be handled by python 
https://github.com/sgl-project/sglang/blob/4b9bf40b2cc860e6f2702173f38d0588265f6525/sgl-router/src/protocols/rerank.rs#L12-L18
